### PR TITLE
Closes #220

### LINKED ISSUE #220: Create dataset loader for n2c2 2014 - Deidentification & Heart Disease
## Adding a Dataset
- **Name:** n2c2 2014 - Deidentification & Heart Disease
- **Description:** *None provided*
- **Task:** NER,DOC_CLASS
- **Paper:** https://pubmed.ncbi.nlm.nih.gov/26225918/
- **Data:** https://portal.dbmi.hms.harvard.edu/projects/n2c2-nlp/
- **License:** DUA-C/NC

### PR DESCRIPTION
**Pull request by jdposada**
Please name your PR after the issue it closes. You can use the following line: "Closes #ISSUE-NUMBER" where you replace the ISSUE-NUMBER with the one corresponding to your dataset.

If the following information is NOT present in the issue, please populate:

- **Name:** *name of the dataset*
- **Description:** *short description of the dataset (or link to social media or blog post)*
- **Paper:** *link to the dataset paper if available*
- **Data:** *link to the online home of the dataset*

### Checkbox

- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `biodatasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_BIGBIO_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `BigBioConfig` for the source schema and one for a bigbio schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_bigbio biodatasets/my_dataset/my_dataset.py`.
- [x] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

```bash
(bigscience-biomedical) root@docker-desktop:/workspaces/biomedical# python -m tests.test_bigbio biodatasets/n2c2_2014/n2c2_2014.py --data_dir /workspaces/biomedical/biodatasets/n2c2_2014/tar_gz
INFO:__main__:args: Namespace(path='biodatasets/n2c2_2014/n2c2_2014.py', schema=None, subset_id=None, data_dir='/workspaces/biomedical/biodatasets/n2c2_2014/tar_gz', use_auth_token=None)
INFO:__main__:self.PATH: biodatasets/n2c2_2014/n2c2_2014.py
INFO:__main__:self.SUBSET_ID: n2c2_2014
INFO:__main__:self.SCHEMA: None
INFO:__main__:self.DATA_DIR: /workspaces/biomedical/biodatasets/n2c2_2014/tar_gz
INFO:__main__:Checking for _SUPPORTED_TASKS ...
INFO:__main__:Found _SUPPORTED_TASKS=[<Tasks.NAMED_ENTITY_RECOGNITION: 'NER'>, <Tasks.TEXT_CLASSIFICATION: 'TXTCLASS'>]
INFO:__main__:_SUPPORTED_TASKS implies _MAPPED_SCHEMAS={'KB', 'TEXT'}
INFO:__main__:schemas_to_check: {'KB', 'TEXT'}
INFO:__main__:Checking load_dataset with config name n2c2_2014_source
WARNING:datasets.builder:Using custom data configuration n2c2_2014_source-57e9df040ed9f011
Downloading and preparing dataset n2c2_2014/n2c2_2014_source to /home/jovyan/.cache/huggingface/datasets/n2c2_2014/n2c2_2014_source-57e9df040ed9f011/1.0.0/1e3e609086987a589ee6db853ac5c55fe73648c31096f144bb0dbbd670cb2f72...
Dataset n2c2_2014 downloaded and prepared to /home/jovyan/.cache/huggingface/datasets/n2c2_2014/n2c2_2014_source-57e9df040ed9f011/1.0.0/1e3e609086987a589ee6db853ac5c55fe73648c31096f144bb0dbbd670cb2f72. Subsequent calls will reuse this data.
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 282.37it/s]
INFO:__main__:Checking load_dataset with config name n2c2_2014_bigbio_kb
WARNING:datasets.builder:Using custom data configuration n2c2_2014_bigbio_kb-57e9df040ed9f011
Downloading and preparing dataset n2c2_2014/n2c2_2014_bigbio_kb to /home/jovyan/.cache/huggingface/datasets/n2c2_2014/n2c2_2014_bigbio_kb-57e9df040ed9f011/1.0.0/1e3e609086987a589ee6db853ac5c55fe73648c31096f144bb0dbbd670cb2f72...
Dataset n2c2_2014 downloaded and prepared to /home/jovyan/.cache/huggingface/datasets/n2c2_2014/n2c2_2014_bigbio_kb-57e9df040ed9f011/1.0.0/1e3e609086987a589ee6db853ac5c55fe73648c31096f144bb0dbbd670cb2f72. Subsequent calls will reuse this data.
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 384.62it/s]
INFO:__main__:Checking load_dataset with config name n2c2_2014_bigbio_text
WARNING:datasets.builder:Using custom data configuration n2c2_2014_bigbio_text-57e9df040ed9f011
Downloading and preparing dataset n2c2_2014/n2c2_2014_bigbio_text to /home/jovyan/.cache/huggingface/datasets/n2c2_2014/n2c2_2014_bigbio_text-57e9df040ed9f011/1.0.0/1e3e609086987a589ee6db853ac5c55fe73648c31096f144bb0dbbd670cb2f72...
Dataset n2c2_2014 downloaded and prepared to /home/jovyan/.cache/huggingface/datasets/n2c2_2014/n2c2_2014_bigbio_text-57e9df040ed9f011/1.0.0/1e3e609086987a589ee6db853ac5c55fe73648c31096f144bb0dbbd670cb2f72. Subsequent calls will reuse this data.
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 514.01it/s]
INFO:__main__:Checking global ID uniqueness
INFO:__main__:Found 12490 unique IDs
INFO:__main__:Gathering schema statistics
INFO:__main__:Gathering schema statistics
train
==========
id: 790
document_id: 790
passages: 790
entities: 17405
normalized: 0
events: 0
coreferences: 0
relations: 0

test
==========
id: 514
document_id: 514
passages: 514
entities: 11462
normalized: 0
events: 0
coreferences: 0
relations: 0

INFO:__main__:Checking if referenced IDs are properly mapped
INFO:__main__:KB ONLY: Checking passage offsets
INFO:__main__:KB ONLY: Checking entity offsets

<SPECIFIC ERRORS ARE HIDDEN> 

There are features with wrong offsets! This is not a hard failure, as it is common for this type of datasets. However, if the error list is long (e.g. >10) you should double check your code. 


INFO:__main__:KB ONLY: Checking event offsets
INFO:__main__:KB ONLY: Checking coref offsets
INFO:__main__:Checking global ID uniqueness
INFO:__main__:Found 514 unique IDs
INFO:__main__:Gathering schema statistics
INFO:__main__:Gathering schema statistics
train
==========
id: 790
document_id: 790
text: 790
labels: 16501

test
==========
id: 514
document_id: 514
text: 514
labels: 10970

.
----------------------------------------------------------------------
Ran 1 test in 11.258s

OK
```

### CODE REVIEW

**Comment by jdposada** [comment]
Closes #220

**Comment by galtay** [comment]
@jdposada this is great and the tests pass on my end but I was wondering if you have the bandwidth to split this into two dataset loaders like was done for `n2c2_2006` ? 
* https://github.com/bigscience-workshop/biomedical/tree/master/biodatasets/n2c2_2006_deid
* https://github.com/bigscience-workshop/biomedical/tree/master/biodatasets/n2c2_2006_smokers

**Comment by jdposada** [comment]
Hi @galtay,

I could do the separation next week, is that ok?

File: biodatasets/n2c2_2014/n2c2_2014.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+            uid = it.count(0)
+            for fname, task in file_names:
+                full_path = os.path.join(data_dir, fname)
+                for x in self._read_tar_gz(full_path):
+                    xml_flag = x["xml_flag"]
+                    if xml_flag:
+                        if task == "track1":
+                            document = self._read_task1_file(file_object=x["file_object"], file_name=x["file_name"])
+                        elif task == "track2":
+                            pass
```
Comment (jdposada) [inline biodatasets/n2c2_2014/n2c2_2014.py:271]: Needs fixing @jdposada

**Comment by galtay** [comment]
> Hi @galtay,
> 
> I could do the separation next week, is that ok?

next week as in the end of the month? that might be a little late ... let's check in this weekend and see where we're at

**Review by galtay — CHANGES_REQUESTED** [review]
split into task1 and task2 dataset loaders

**Comment by jason-fries** [comment]
Hey @jdposada 
Just checking in on this PR -- do you have bandwidth to make the requested changes?

**Comment by ruisi-su** [comment]
Hi @jdposada, just checking on the timeline. Any chance that this could be done before the end of next week? Thanks in advance!

**Comment by ruisi-su** [comment]
Hi @jdposada, thanks for your hard work in adding this dataset! Any chance that this could be PR-ready by next week? We really need to add this dataset ASAP. Please tell us if you would not be able to have the bandwidth to get this ready. We will need to re-assign the issue. I wanted to check with you first since you have already worked on this for quite a while. Thanks again!

**Comment by jdposada** [comment]
Hi @ruisi-su , working on it, I understand the urgency and surely will be able to make the PR. I know I missed my past deadline but will firmly commit to this one. Thanks

**Comment by jdposada** [comment]
I created a different pull request since I am splitting the datasets. The new pull request for deid is here. The risk stratification task is coming

https://github.com/bigscience-workshop/biomedical/pull/644

**Comment by jdposada** [comment]
Hi @ruisi-su 

I created the second pull request for risk factors here. Let me know if I am missing anything. Thanks

https://github.com/bigscience-workshop/biomedical/pull/646

**Comment by ruisi-su** [comment]
This PR is closed as the subtasks are addressed in #644 and #646

**Pull request closed without merging**